### DISCUSSION TITLE
FDK Cuda Reconstruction Parameters

### DISCUSSION THREAD

**Discussion by ZacharyEBirch** [astra-toolbox/astra-toolbox#456] [Q&A] [answered]
Is it possible to specify which slices you want to reconstruct when using FDK Cuda or can you only do the entire set?

**Comment by askorikov** [comment]
Hi! You can specify a window for a given 3D geometry, if that's what you mean: https://astra-toolbox.com/docs/geom3d.html#volume-geometries

**Reply by ZacharyEBirch** [reply]
Thanks for the response!

I meant more if I could do reconstruction skips to reduce the total volume. I am only trying to do skips in one direction and changing the projection/volume geometry I believe would require treating them as anisotropic pixels which FDK doesn't allow. 

I hope that makes sense.

**Reply by askorikov** [reply]
Yes, that is not really supported at the moment. What you can try is reconstructing each slice one-by-one in a for loop, but that will most likely have serious performance implications (could still be doable if your reconstruction is not too big).

**Reply by ZacharyEBirch** [reply]
Ah, okay. I only need a few slices for my application so a for-loop is not out of the question. Is there a function that allows extracting a slice from an FDK-based 3D volume?

**Reply by askorikov** [reply]
Creating an appropriately positioned volume geometry with a single slice would be the most direct implementation of that I think.

**Reply by ZacharyEBirch** [reply]
Ah so we implement fanbeam geometry on each slice where the strip is at the relevant height relative to the source?

**Reply by askorikov** [reply, marked as answer]
I guess strictly speaking that wouldn't be fanbeam geometry because the source is not at the same height as the reconstructed slice (for this reason, you will still need the entire input projection data, since many rows of the detector will influence the reconstruction). But now I'm thinking I probably didn't properly get the situation you have in mind. Anyway, here is a code snippet showing what I meant:

```matlab
n = 1000;

proj_geom = astra_create_proj_geom('cone', 1.0, 1.0, n, n, linspace(0, 2*pi, 360), 3000, 3000);
sino_id = astra_mex_data3d('create', '-sino', proj_geom, 1.0);
cfg = astra_struct('FDK_CUDA');
cfg.ProjectionDataId = sino_id;

% Reconstruct every 10th slice
vol = zeros(n, n, 1, 'single');
for slice_n = 1:10:n
    slice_height = -n/2 + slice_n;
    vol_geom = astra_create_vol_geom(n, n, 1, -n/2, n/2, -n/2, n/2, ...
                                     slice_height - 1/2, slice_height + 1/2);
    rec_id = astra_mex_data3d('create', '-vol', vol_geom);
    cfg.ReconstructionDataId = rec_id;
    alg_id = astra_mex_algorithm('create', cfg);
    astra_mex_algorithm('run', alg_id);
    vol(:, :, end+1) = astra_mex_data3d('get', rec_id);
end
```

**Reply by ZacharyEBirch** [reply]
This works quite well, thanks!

**Reply by ZacharyEBirch** [reply]
Actually, there is one issue. 

For the vol_geom it only seems to work within a certain range of slice_height values. I can attach my values for reference.

```matlab
proj_geom = astra_create_proj_geom('cone', 0.008, 0.008, 4096, 4096, angles, 65, 195);
vol_geom = astra_create_vol_geom(4096, 4096, 1, -4.096, 4.096, -4.096, 4.096, slice_height-0.001, slice_height+0.001);
```

This only seems to be valid when -0.499 < slice_height < 0.499 otherwise it throws the anisotropic FDK error. Not sure why.

**Reply by ZacharyEBirch** [reply]
I ended up cheesing it by normalizing my geometry and pixel pitch so that my volume min/max matched my pixel values (-2048, 2048) and it worked
